### PR TITLE
Enable AI

### PR DESCRIPTION
**Pull request by juniwang**
Enable AI for events:
- agent ssh launched 
- agent provisioned
- agent shutdown
- agent deprovisioned
- azure rest call

Build may fail since azure-commons plugin is not yet released to Jenkins plugin hub.

### CODE REVIEW

File: src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java [hunk 1]
```diff
@@ -238,6 +228,11 @@ public void onClosed(Channel channel, IOException cause) {
             // state to the default for the node.
             agent.clearCleanUpAction();
             successful = true;
+
+            // send AI event
+            final Map<String, String> properties = new HashMap<>();
+            properties.put("OSType", agent.getOsType().toString());
+            AzureVMAgentPlugin.sendEvent(Constants.AI_VM_AGENT, "SSHLaunch", properties);
         } catch (Exception e) {
             LOGGER.log(Level.INFO, "AzureVMAgentSSHLauncher: launch: got exception ", e);
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: Also send event if there is exception

File: src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java [hunk 2]
```diff
@@ -841,6 +826,15 @@ private void handleFailure(AzureVMAgentTemplate template, String vmName, Excepti
...
         LOGGER.log(Level.INFO,
                 "AzureVMCloud: provision: asynchronous provision finished, returning {0} planned node(s)", plannedNodes.size());
+
+        final Map<String, String> properties = new HashMap<>();
+        properties.put("NumberOfAgents", String.valueOf(plannedNodes.size()));
+        properties.put(AppInsightsConstants.AZURE_SUBSCRIPTION_ID,
+                template.getAzureCloud().getServicePrincipal().getSubscriptionId());
+        properties.put(AppInsightsConstants.AZURE_LOCATION, template.getLocation());
+        properties.put("AgentOS", template.getOsType());
+        AzureVMAgentPlugin.sendEvent(Constants.AI_VM_AGENT, "Provision", properties);
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java]: Also send event if provision failed.

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java [hunk 3]
```diff
@@ -462,6 +462,10 @@ public void deprovision(Localizable reason) throws Exception {
         }
 
         Jenkins.getInstance().removeNode(this);
+
+        final Map<String, String> properties = new HashMap<>();
+        properties.put("Reason", reason.toString());
+        AzureVMAgentPlugin.sendEvent(Constants.AI_VM_AGENT, "deprovision", properties);
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java]: **D**eprovision

File: src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java (hunk 2 above)
Comment (juniwang) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java]: It's  a long method with many `try{..}catch{...}` blocks.  Will try my best to find some place to send an event but cannot make sure it covers all branches.

**Comment by juniwang** [comment]
Don't merge, still have some issues.

**Comment by juniwang** [comment]
issues resolved. Plz review again.

**Pull request merged**